File: 2016/day04_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any

type Room = tuple[str, int, str]


def solution(content: list[str]) -> Any:
    return "\n".join(
        str(num)
        for name, num, check in (parse(line) for line in content)
        if check == checksum(name) and "north" in decrypt(name, num)
    )


def decrypt(name: str, key: int):
    return "".join(rotate(char, key) for char in name)


def rotate(char: str, key: int):
    if char == "-":
        return " "
    return chr(ord("a") + ((ord(char) - ord("a") + key) % 26))


def checksum(name: str):
    def helper(count: dict[str, int], line: list[str]):
        if not line:
            return "".join(
                key
                for key, _ in sorted(
                    sorted(count.items(), key=lambda item: item[0]),
                    key=lambda item: item[1],
                    reverse=True,
                )
            )[:5]
        char, *rest = line
        count[char] = 1 + count.get(char, 0)
        return helper(count, rest)

    return helper({}, list(name.replace("-", "")))


def parse(line: str) -> Room:
    match = re.match(r"([a-z\-]+)(\d+)\[([a-z]+)\]", line)
    if match is None:
        raise ValueError(f"Invalid line: {line}")
    return (match.group(1), int(match.group(2)), match.group(3))


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2016/day04-2.ts
type Room = {
  name: string;
  code: number;
  checksum: string;
};

function solution(content: string) {
  return content
    .split("\n")
    .map(parse)
    .filter((room) => checksum(room.name) === room.checksum)
    .map((room) => ({
      ...room,
      name: decrypt(room.name, room.code),
    }))
    .filter((room) => room.name.includes("north"))
    .at(0)?.code;
}

function decrypt(name: string, key: number) {
  return name
    .split("")
    .map((char) => rotate(char, key))
    .join("");
}

function rotate(char: string, key: number) {
  if (char === "-") return " ";
  return String.fromCharCode(
    "a".charCodeAt(0) + ((char.charCodeAt(0) - "a".charCodeAt(0) + key) % 26)
  );
}

function checksum(name: string) {
  function helper(count: Record<string, number>, line: string) {
    if (line.length === 0) {
      return Object.entries(count)
        .toSorted(
          ([charA], [charB]) => charA.charCodeAt(0) - charB.charCodeAt(0)
        )
        .toSorted(([_A, countA], [_B, countB]) => countB - countA)
        .map(([char]) => char)
        .join("")
        .slice(0, 5);
    }
    const char = line.at(0)!;
    const rest = line.slice(1);
    count[char] = 1 + (count[char] ?? 0);
    return helper(count, rest);
  }
  return helper({}, name.replaceAll("-", ""));
}

function parse(line: string): Room {
  const match = line.match(/([a-z\-]+)(\d+)\[([a-z]+)\]/);
  if (!match) throw new Error(`Invalid line: ${line}`);
  return {
    checksum: match.at(3)!,
    code: Number(match.at(2)),
    name: match.at(1)!,
  };
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
